src/main/java/rockset/RocksetSinkTask.java

```diff
@@ -71,164 +54,98 @@ public void start(Map<String, String> settings) {
+      executorService.shutdownNow();
     }
     log.info("Stopped Rockset Kafka Connect Plugin");
   }
 
   // open() will be called for a partition before any put() is called for it
   @Override
   public void open(Collection<TopicPartition> partitions) {
-    log.debug(String.format("Opening %d partitions: %s", partitions.size(), partitions));
-    partitions.forEach(
-        tp -> {
-          Preconditions.checkState(!futureMap.containsKey(tp));
-          futureMap.put(tp, new ArrayList<>());
-        });
+    log.debug("Opening {} partitions: {}", partitions.size(), partitions);
   }
 
   @Override
   public void close(Collection<TopicPartition> partitions) {
-    log.debug(String.format("Closing %d partitions: %s", partitions.size(), partitions));
-    partitions.forEach(
-        tp -> {
-          Preconditions.checkState(futureMap.containsKey(tp));
-          futureMap.remove(tp);
-        });
+    log.debug("Closing {} partitions: {}", partitions.size(), partitions);
   }
 
-  // put() doesn't need to block until the writes complete, that is what flush() is for
   @Override
   public void put(Collection<SinkRecord> records) {
-    if (records.size() == 0) {
-      log.debug("zero records in put call, returning");
-      return;
-    }
-
-    groupRecordsByTopicPartition(records)
-        .forEach(
-            (tp, recordBatch) -> {
-              try {
-                RetriableTask task =
-                    new RetriableTask(
-                        taskExecutorService,
-                        retryExecutorService,
-                        () -> addDocs(tp.topic(), recordBatch));
-
-                // this should only block if all the threads are busy
-                taskExecutorService.submit(task);
-
-                Preconditions.checkState(futureMap.containsKey(tp));
-                futureMap.get(tp).add(task);
-              } catch (InterruptedException e) {
-                throw new ConnectException("Failed to put records", e);
+    Map<TopicPartition, Future<?>> futures =
+        records.stream()
+            .collect(Collectors.groupingBy(r -> new TopicPartition(r.topic(), r.kafkaPartition())))
+            .entrySet()
+            .stream()
+            .collect(
+                Collectors.toMap(
+                    Map.Entry::getKey,
+                    e ->
+                        executorService.submit(
+                            () ->
+                                requestWrapper.addDoc(
+                                    e.getKey().topic(),
+                                    e.getValue(),
+                                    recordParser,
+                                    config.getRocksetBatchSize()))));
+    futures.forEach(
+        (tp, f) -> {
+          try {
+            f.get();
+          } catch (Exception e) {
+            // Stop all tasks still executing since this put() will be retried anyway
+            futures.values().forEach(g -> g.cancel(true));
```

Comment:
for my own understanding - does cancelling cause other tasks to throw `ConnectException` even if the current future throws `RetriableException`? Does that halt from sending more requests?

Reply:
This cancels the thread which will throw an InterruptedException, but that's not handled anywhere. We also don't process the futures for any of the other writes once we see one of them has failed, so even if they did throw an exception it wouldn't matter. We're doing here is propagating the first error that we see. So if there are a combination of RetriableExceptions and ConnectExceptions across all the futures, just the first one will win.

Reply:
can we add a unit test to cover such scenario?